Correct CAI Implementation and Multiple Genetic Codes

Benjamin-Lee opened the pull request:
Fixes an incorrect implementation of the CAI formula as defined by Sharp and Li. They state "that if a certain codon is never used in the reference set then the CAI for any other gene in which that codon appears becomes zero. To overcome this problem we assign a value of 0.5 to any [codon] that would otherwise be zero." Also adds support for multiple genetic codes.

peterjc commented:
I've had a quick look at your changes, but unfortunately it looks like your editor is setup to indent with 8 spaces rather than the Python norm of 4 spaces (see PEP8, https://www.python.org/dev/peps/pep-0008/#indentation ), which Biopython tries to follow.

This makes it much harder to interpret your changes. Could you the indentation and then update the pull request? Thanks!

Also, based on the TravisCI testing, you need to look at why ``test_CodonUsage.py`` has stopped working (it looks like an accidental API change rather than a deliberate numerical result changing).

Benjamin-Lee commented:
I just fixed the PEP8 errors and fixed the API to ensure back compatibility. It's now passing TravisCI's tests so the current commit should be ready to take a look at.

Thanks,
Benjamin

peterjc commented:
OK, this needs clarification:

> Copyright 2016 Sandia Corporation. Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains certain rights in this software.

What rights? Also this would make be very nervous about hampering our planned move to adopt the standard 3-clause BSD licence in future http://lists.open-bio.org/pipermail/biopython-dev/2015-November/021151.html

Benjamin-Lee commented:
Actually, I was informed today that the intellectual property department has said that Sandia is not going to be retaining rights or copyright to this code, so I removed the line. Sorry for the delay while I clarified it with them.

peterjc commented:
Thanks for clarifying @Benjamin-Lee. On this point, would you allow us to re-licensing your contributions under the 3-clause BSD as outlined in #898 ?

Benjamin-Lee commented:
Yes. I just double checked and the 3-clause BSD will be fine.

peterjc commented:
@Benjamin-Lee Thanks for sorting out the licensing. I'll probably squash this into one commit when it's ready to merge - the overall diff it much easier to read now, thank you!

Looking at the API, you've added a new method ``change_translation_table`` but might it be simpler just to take the table as a ``__init__`` argument (defaulting to table 1 for backward compatibility)? At the same time you could remove the hard-coded ``SynonymousCodons`` dictionary by building it at run time from ``Bio.Data`` for the chosen genetic code.

@Iddo (since you handled the original inclusion of ``Bio.SeqUtils.CodonUsage`` from Yair Benita) and @sbassi (since you've done a lot of work in ``Bio.SeqUtils``): any comments?

iddo commented:
@peterjc You mentioned me by mistake

peterjc commented:
@iddo my apologies, I must have been too quick with the auto-completion and got the wrong Iddo.

@idoerg - you checked in the first version of ``Bio.SeqUtils.CodonUsage`` https://github.com/biopython/biopython/commit/19f997e2fb9e68f0672b8cd735150bcb30ba53b9 - do you have any comments on this pull request (e.g. genetic codes)

Benjamin-Lee commented:
@peterjc I updated the code to change the dictionary to be pulled from `Bio.Data` but it's now saying that the build failed. I think it might be due to the fact that the dictionary is now in a different order. I'm going to look into it some more but wasn't sure what your take on it is now.

peterjc commented:
There's a minor style issue to fix, spotted by the pep8 command line tool"

```
Bio/SeqUtils/CodonUsage.py:39:1: W293 blank line contains whitespace
```

As to the test failure, ``test_CodonUsage.py`` is an old print-and-compare test. I doubt is is the dictionary order as that shouldn't matter - the ``print_index`` function sorts the keys:

```
    def print_index(self):
        """Prints out the index used."""
        for i in sorted(self.index):
            print("%s\t%.3f" % (i, self.index[i]))
```

Try running ``python test_CodonUsage.py`` directly, and/or ``python run_tests.py -g test_CodonUsage.py`` and ``git diff output/test_CodinUsage`` to see what is going on. There is some background information on the print-and-compare tests in the Biopython Tutorial testing chapter which might help: http://biopython.org/DIST/docs/tutorial/Tutorial.html

(I'm currently on leave and haven't tried your changes here yet)

peterjc commented:
The tests are still failing, either something about the sorting or perhaps more likely your changes are adding entries for previously silent codons?

```
======================================================================
ERROR: runTest (__main__.ComparisonTestCase)
test_CodonUsage
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_tests.py", line 348, in runTest
    % (repr(output_line), repr(expected_line)))
ValueError: 
Output  : 'TAC\t1.000'
Expected: 'TAA\t1.000'
----------------------------------------------------------------------
```

Did you try ``python run_tests.py -g test_CodonUsage.py`` (in ``Tests/`` folder) and looking at how ``Tests/output/test_CodonUsage`` changed (plain text file of expected output from ``test_CodonUsage.py``)?

peterjc commented:
Did you try updating ``Tests/output/test_CodonUsage`` here?

mdehoon commented:
This was fixed in #4085 (fixing the incorrect implementation, and allowing for different genetic codes).
